arrow/arrow.py

```diff
@@ -1,146 +1,112 @@
+        '''
+        This invokes the Obsidian C module with the stoichiometry, reaction rates and
+        a variety of derived values. Once constructed, this can be invoked by calling
+        `evolve` with a duration and initial state, since the stoichiometry will be
+        shared among all calls.
+
+        There are four derived values, each of which is a list of variable length
+        lists. In order to pass this into C, these nested lists are flattened and two
+        associated lists are constructed: one to hold the indexes for each sublist and
+        one to hold the lengths of these sublists. So, to access one of the sublists
+        at index `i`, you can find the start index of the sublist inside of `_flat`
+        with `_indexes[i]`, then iterate through it for the number of elements given
+        by `_lengths[i]`. Strictly speaking this could have been implemented with only
+        the lengths (or the indexes, they are in a way duals of each other), but that
+        would require additional computation. Since this effort is motivated by
+        performance, in order to trade time for space the most expedient approach is
+        chosen, even if it is more verbose.
+        '''
+
+        self.stoichiometry = stoichiometry
         self.rates = rates
 
-        reactants, reactant_stoichiometries = derive_reactants(stoichiometric_matrix)
-
-        self.reactants = reactants
-        self.reactant_stoichiometries = reactant_stoichiometries
-        self.dependencies = calculate_dependencies(stoichiometric_matrix)
+        reactants, reactions, substrates = derive_reactants(stoichiometry)
+        dependencies = calculate_dependencies(stoichiometry)
 
-    def step(self, state):
-        return step(self.stoichiometric_matrix, self.rates, state)
+        self.reactants_flat, self.reactants_lengths, self.reactants_indexes = flat_indexes(reactants)
+        self.reactions_flat = np.array(flatten(reactions))
+        self.dependencies_flat, self.dependencies_lengths, self.dependencies_indexes = flat_indexes(
+            dependencies)
+        self.substrates_flat, self.substrates_lengths, self.substrates_indexes = flat_indexes(substrates)
 
-    def evolve(self, state, duration):
-        return evolve(
-            self.stoichiometric_matrix,
+        self.obsidian = obsidian.obsidian(
+            self.stoichiometry,
             self.rates,
-            state,
-            duration,
-            reactants=self.reactants,
-            reactant_stoichiometries=self.reactant_stoichiometries,
-            dependencies=self.dependencies)
+            self.reactants_lengths,
+            self.reactants_indexes,
+            self.reactants_flat,
+            self.reactions_flat,
+            self.dependencies_lengths,
+            self.dependencies_indexes,
+            self.dependencies_flat,
+            self.substrates_lengths,
+            self.substrates_indexes,
+            self.substrates_flat)
```

Comment:
Yeah I thought about this but wanted to keep things simple implementation-wise, though a struct for the nested arrays could be warranted. The main thing that dissuaded me from this is how `reactants` and `reactions` shared the same `_lengths` and `_indexes` and it seemed honoring this would introduce more complexity than simply passing everything in. 

I can look into the cdef/cython thing, though at this point that sounds like a new implementation, and it is not guaranteed to not introduce other slowdowns we are not anticipating (my attempts with cython on this problem before showed minimal speedups, though this avenue was not explored exhaustively).

Reply:
Yeah the approach there is to let Cython do a thin glue layer and leave the work to C code. I know that's a big change for all the obsidianmodule.c code. The result ought to be a lot simpler and more maintainable...

Reply:
Yeah that does sound simpler. I think since this is working I'm going to roll with it and if I ever need to come back here and modify anything instead of wrestling with the module code again I'll try out the cython-as-glue approach.